File: src/main/java/cn/chahuyun/authorize/HuYanAuthorize.java
package cn.chahuyun.authorize;

import cn.chahuyun.authorize.command.AuthorizeCommand;
import cn.chahuyun.authorize.config.AuthorizeConfig;
import cn.chahuyun.authorize.utils.HibernateUtil;
import net.mamoe.mirai.console.command.CommandManager;
import net.mamoe.mirai.console.plugin.jvm.JavaPlugin;
import net.mamoe.mirai.console.plugin.jvm.JvmPluginDescriptionBuilder;
import net.mamoe.mirai.utils.MiraiLogger;
import xyz.cssxsh.mirai.hibernate.MiraiHibernateConfiguration;

/**
 * @author Moyuyanli
 */
public final class HuYanAuthorize extends JavaPlugin {

    /**
     * 插件本身唯一实例
     */
    public static final HuYanAuthorize INSTANCE = new HuYanAuthorize();
    /**
     * 全局版本
     */
    public static final String VERSION = "1.0.7";
    /**
     * 日志
     */
    public static final MiraiLogger LOGGER = INSTANCE.getLogger();

    private HuYanAuthorize() {
        super(new JvmPluginDescriptionBuilder("cn.chahuyun.HuYanAuthorize", VERSION)
                .name("HuYanAuthorize")
                .author("Moyuyanli")
                .info("壶言权限管理")
                .dependsOn("xyz.cssxsh.mirai.plugin.mirai-hibernate-plugin", false)
                .build());
    }

    @Override
    public void onEnable() {
        //加载配置
        reloadPluginConfig(AuthorizeConfig.INSTANCE);
        //加载指令
        CommandManager.INSTANCE.registerCommand(AuthorizeCommand.INSTANCE, true);
        //加载配置
        MiraiHibernateConfiguration configuration = new MiraiHibernateConfiguration(this);
        //初始化插件数据库
        HibernateUtil.init(configuration);
        //添加本插件的注册消息包信息
        PermissionServer instance = PermissionServer.getInstance();
        instance.init(INSTANCE, "cn.chahuyun.authorize.manager");
        LOGGER.info("HuYanAuthorize plugin loaded!");
    }

    @Override
    public void onDisable() {
        LOGGER.info("HuYanAuthorize plugin uninstall!");
    }
}

File: src/main/kotlin/cn/chahuyun/authorize/HuYanAuthorize.kt
package cn.chahuyun.authorize

import cn.chahuyun.authorize.command.AuthorizeCommand
import cn.chahuyun.authorize.config.AuthorizeConfig
import cn.chahuyun.authorize.constant.AuthPerm
import cn.chahuyun.authorize.entity.Perm
import cn.chahuyun.authorize.utils.PermUtil
import cn.chahuyun.authorize.utils.UserUtil
import net.mamoe.mirai.console.command.CommandManager.INSTANCE.register
import net.mamoe.mirai.console.plugin.jvm.JvmPluginDescription
import net.mamoe.mirai.console.plugin.jvm.KotlinPlugin

/**
 * @author Moyuyanli
 */
object HuYanAuthorize : KotlinPlugin(
    JvmPluginDescription(
        id = "cn.chahuyun.HuYanAuthorize",
        version = BuildConstants.VERSION,
        name = "HuYanAuthorize"
    ) {
        author("Moyuyanli")
        info("壶言权限管理")
    }
) {

    val log by lazy { logger }

    override fun onEnable() {
        // 加载配置
        AuthorizeConfig.reload()
        // 加载指令
        AuthorizeCommand.register()
        // 初始化插件数据库
        DataManager.init(this)
        // 注册本插件的权限
        AuthorizeServer.registerPermissionsInternal(
            this,
            Perm(AuthPerm.OWNER, "主人权限"),
            Perm(AuthPerm.ADMIN, "管理员权限")
        )
        // 添加本插件的注册消息
        AuthorizeServer.registerEvents(this, "cn.chahuyun.authorize.manager")
        // 尝试同步主人
        syncOwner()
        logger.info("HuYanAuthorize plugin loaded!")
    }

    override fun onDisable() {
        logger.info("HuYanAuthorize plugin uninstalled!")
    }

    /**
     * 同步主人信息
     * 确保配置文件和数据库中的“主人”权限组一致
     */
    private fun syncOwner() {
        val configOwner = AuthorizeConfig.owner
        val permGroup = PermUtil.takePermGroupByName("主人")

        if (configOwner == 123456L) {
            // 如果配置是默认值，尝试从数据库权限组中找回主人
            val firstOwner = permGroup.users.firstOrNull()?.userId
            if (firstOwner != null) {
                AuthorizeConfig.owner = firstOwner
                log.info("从数据库同步主人信息: $firstOwner")
            } else {
                log.warning("当前未设置主人（默认 123456），请尽快设置主人权限！")
            }
        } else {
            // 如果配置已经指定了主人，确保他在权限组里
            val globalUser = UserUtil.globalUser(configOwner)
            if (permGroup.users.none { it.userId == configOwner }) {
                permGroup.addUser(globalUser)
                log.info("已将配置文件中的主人 $configOwner 同步至权限组")
            }
        }
    }

    /**
     * 获取主人
     */
    fun getOwner(): Long {
        return AuthorizeConfig.owner
    }

}